:couple: BN254 EC Pairing implementation under Boojum constraint system

ZamDimon opened the pull request:
# What ❔

This pull request primarily focuses on EC pairing implementation under the constraint system. That is, we implement the function $e: E(\mathbb{F}\_p)[r] \times E\'(\mathbb{F}\_{p^2})[r] \to \mathbb{F}\_{p^{12}}^{\star}$

To implement that, we implement the following functionality:
- Tower extension $\mathbb{F}\_{p^2} \mapsto \mathbb{F}\_{p^6} \mapsto \mathbb{F}\_{p^{12}}$ under the constraint system.
- Twisted elliptic curve $E'(\mathbb{F}\_{p^2})$ support.
- Consequently, we implement the `NonNativeField<F, T>` trait over $\mathbb{F}\_{p^2}$ extension and basic arithmetic for all of the extensions.
- We implement the essential components of the elliptic curve pairing:
  - Line and tangent functions evaluation.
  - Basic wnaf-based Miller loop implementation.
  - Basic final exponentiation

## Why ❔

1. Currently, no EC pairing is implemented under the constraint system.
2. Field extensions are implemented outside the constraint system (see [here](https://github.com/matter-labs/pairing/tree/master/src/bn256)); however, no such functionality is available under the constraint system.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

jules commented on src/gadgets/curves/bn256/ec_mul.rs:
```diff
@@ -0,0 +1,440 @@
+        CS: ConstraintSystem<F>,
+    {
+        // Defining constants under the constraint system
+        let pow_2_128 = UInt512::allocated_constant(cs, (U256([0, 0, 1, 0]), U256::zero()));
```
As it is used for the bounds checks of the scalars, I believe it should correspond to the max decomposition value, which is ceil(sqrt(n)) and should be lower than 2^128. You can double check it, but I believe this max decomposition value should be:

```rust
let pow_2_128 = UInt512::allocated_constant(
    cs,
    (
        U256([14338140286821604122, 8020209761171036668, 0, 0]),
        U256::zero(),
    ),
);
```

jules replied:
hmm, are we sure this is correct as well then?

jules commented on src/gadgets/curves/bn256/ec_mul.rs:
```diff
@@ -0,0 +1,440 @@
+        let low_pow_2_128 = pow_2_128.to_low();
+        
+        // Selecting between k1 and -k1 in Fq
+        let (_, k1_out_of_range) = k1_u256.overflowing_sub(cs, &low_pow_2_128);
+        let k1_negated = k1.negated(cs);
+        let k1 = <BN256ScalarNNField<F> as NonNativeField<F, BN256Fr>>::conditionally_select(
+            cs,
+            k1_out_of_range,
+            &k1,
+            &k1_negated,
+        );
```
there is a small error here when k1 == max_decomp_value - it will get negated when it shouldn't. a fix would be:

```rust
        let (_, k1_out_of_range) = low_pow_2_128.overflowing_sub(cs, &k1_u256);
        let k1_negated = k1.negated(cs);
        let k1 = <BN256ScalarNNField<F> as NonNativeField<F, BN256Fr>>::conditionally_select(
            cs,
            k1_out_of_range,
            &k1_negated,
            &k1,
        );
```

same goes for k2, below here

jules replied:
it is also a bit more clear in this sense, since k1 is only out of range is if it above the decomp value, but in the current case it is considered out of range if its smaller than the decomp value

ZamDimon replied:
Yeah, thank you! I could not figure out why I had to negate the booleans further to make it work; now, it makes sense.

Fixed!

jules replied:
i'll still recommend to use the example here as it removes 2 negations from the circuit

jules commented on src/gadgets/curves/bn256/ec_mul.rs:
```diff
@@ -0,0 +1,440 @@
+        Self {
+            k1,
+            k2,
+            k1_was_negated: k1_out_of_range.negated(cs),
+            k2_was_negated: k2_out_of_range.negated(cs),
```
the previous comment's update would also mean you need to remove the negation here

jules commented on src/gadgets/curves/bn256/ec_pairing.rs:
```diff
@@ -0,0 +1,303 @@
+            let mut line_fn = evaluation.at_line(cs, &mut r, &mut q, p);
+            f1 = Self::mul_f12_by_line_fn(cs, &mut f1, &mut line_fn);
+
+            let qx = q.x.clone();
+            let qy = q.y.clone();
+            r = r.add_mixed(cs, &mut (qx, qy));
```
would there be any case here where `qz` isn't 1 and so we'd need to reduce to affine first?

ZamDimon replied:
Since the $Q$ point does not change during the loop, as long as the inputted $Q$ is "normalized" ($Z=1$), I believe we are safe to assume $Z=1$.

However, I added the enforcement on `q.z` to be one at the beginning of the function; thanks for noticing that!

jules commented on src/gadgets/curves/bn256/ec_pairing.rs:
```diff
@@ -0,0 +1,303 @@
+}
+
+/// This function computes the pairing function for the BN256 curve.
+pub fn ec_mul<F, CS>(
```
nit: i think you want to name this `ec_pairing` :)

ZamDimon replied:
Haha, yeah, thank you for noticing :)

jules commented on src/gadgets/curves/zeroable_affine/mod.rs:
```diff
@@ -1,73 +1,277 @@
+            .flat_map(|byte| (0..8).rev().map(move |i| (byte >> i) & 1 == 1))
+            .collect::<Vec<_>>();
+
+        for bit in scalar_bits {
+            if bit {
+                result = result.add(cs, &mut temp);
+            }
+            temp.double(cs);
```
this will result in another non-homogenous circuit structure - the addition should happen through conditional select

ZamDimon replied:
Oh, I did not even mention that we have not removed the `zeroable_affine` implementation. Updated it with the `conditionally_select`, but I believe we will not use this part further.

jules commented on src/gadgets/curves/zeroable_affine/mod.rs:
```diff
@@ -1,73 +1,277 @@
+        CS: ConstraintSystem<F>,
+    {
+        // If x's are not the same, adding via unequal_x method
+        let same_x = self.same_x(cs, other);
+        let boolean_false = Boolean::allocated_constant(cs, false);
+        if same_x == boolean_false {
+            return self.add_unequal_x(cs, other);
+        }
+
+        // If y's are the same, doubling the point
+        let same_y = self.same_y(cs, other).negated(cs);
+        if same_y == boolean_false {
+            return self.double(cs);
+        }
```
also non-homogenous, should be done with conditional select gates

ZamDimon replied:
Just removed the function as of now; we probably will not need it :) But we can safely add it further

jules commented on src/gadgets/tower_extension/fq6.rs:
```diff
@@ -0,0 +1,345 @@
+        let is_c0_zero = self.c0.is_zero(cs);
+        let is_c1_zero = self.c1.is_zero(cs);
+        let is_c2_zero = self.c2.is_zero(cs);
+        is_c0_zero.and(cs, is_c1_zero).and(cs, is_c2_zero)
```
`Boolean::multi_and` may work better here, I think its marginally cheaper

ZamDimon replied:
Yeah, besides, this portion looks much more natural now!

jules commented on src/gadgets/curves/bn256/ec_add.rs:
```diff
@@ -0,0 +1,18 @@
+pub fn projective_add<F, CS>(
+    cs: &mut CS,
+    point_1: &mut BN256SWProjectivePoint<F>,
+    point_2: BN256SWProjectivePoint<F>,
```
since we already get X and Y coordinates passed through in the ecadd call, it seems superfluous to cast point 2 to projective and then clone the points out. we should just pass a tuple of coords

jules commented on src/gadgets/curves/bn256/ec_mul.rs:
```diff
@@ -0,0 +1,440 @@
+// -- GLV constants --
+
+// Decomposition scalars can be a little more than 2^128 in practice, so we use 33 chunks of width 4 bits
+const MAX_DECOMPOSITION_VALUE: U256 = U256([u64::MAX, u64::MAX, 0x0f, 0]);
```
this seems unused and best to remove to avoid confusion i think

jules commented on src/gadgets/curves/bn256/ec_pairing.rs:
```diff
@@ -0,0 +1,556 @@
+    {
+        // 1. tmp0 <- X_Q^2; 2. tmp1 <- Y_Q^2; 3. tmp2 <- tmp1^2;
+        let mut tmp0 = q.x.square(cs);
+        tmp0.normalize(cs);
```
do we need this many calls to normalize? i think the nonnative field gadget can take a few overflows before needing to reduce again right?

NikitaMasych replied:
Hey, @jules, the excessive amount of normalization calls is reasoned by the desire to conduct extra check that pairing has no problems. Currently, I am trying to reduce this amount, meanwhile guaranteeing that implementation is still 100% correct. For example, I removed all normalization calls from the doubling step, and all the tests still pass. The issue which I am worried about is that there won't be such an input, which would require normalization, which is not needed for another input. Basically, static checks that the implementations is legit, agnostic of the input. So, in a nutshell, the question is how to understand where to put normalization calls - is it achievable by experimenting in comp-time?

jules replied:
hey @NikitaMasych ! you can ensure that the input passed to the function is reduced beforehand by calling `enforce_reduced`. this should guarantee that your input can always take the exact same amount of overflows before giving an error. hope that helps!

NikitaMasych commented:
Hey, @jules, I have added fixes for the above notes from you, though since we moved BN254-specific functionality to era-zkevm_circuits, you can check the changes in [conjugated PR](https://github.com/matter-labs/era-zkevm_circuits/pull/32)

ZamDimon commented:
Since we moved the BN254-specific implementations outside this repo (namely, we moved them [`here`](https://github.com/matter-labs/era-zkevm_circuits/pull/32)), we rename this pull request to include information about gadgets only. We also close other pull requests not related to gadgets.

jules commented on src/gadgets/non_native_field/implementations/implementation_u16.rs:
```diff
@@ -1032,14 +1035,106 @@ impl<F: SmallField, T: pairing::ff::PrimeField, const N: usize> CSAllocatable<F>
 
-#[derive(Derivative)]
-#[derivative(Clone, Copy, Debug, Hash)]
+#[derive(Derivative, Serialize, PartialEq)]
```
what was the reason for adding serialize functionality to this type?

NikitaMasych replied:
In circuits for `ec_pairing` we use internal FSM state with `Fq12` as intermediate value element, thus it was needed to add it here because otherwise it wouldn't get compiled

jules commented on src/gadgets/tower_extension/algebraic_torus.rs:
```diff
@@ -0,0 +1,299 @@
+        let mut result = Self::compress::<CS, false>(cs, &mut one);
+        let mut found_one = false;
+
+        for i in BitIterator::new(exponent) {
+            if found_one {
+                result = result.square::<CS, SAFE>(cs);
+            } else {
+                found_one = i;
+            }
+
+            if i {
+                result = result.mul::<CS, SAFE>(cs, self);
+            }
```
this is non-homogenous and needs to be updated to use conditional selection gates

jules commented on src/gadgets/tower_extension/fq2.rs:
```diff
@@ -0,0 +1,633 @@
+    where
+        CS: ConstraintSystem<F>,
+    {
+        if power % 2 == 0 {
+            return self.clone();
+        }
```
also non-homogenous in case the power can variate between circuits

jules commented on src/gadgets/tower_extension/fq12.rs:
```diff
@@ -0,0 +1,686 @@
+        let mut result = Self::one(cs, self.c0.c0.get_params());
+        let mut found_one = false;
+
+        for i in BitIterator::new(exponent) {
+            if found_one {
+                result = result.square(cs);
+            } else {
+                found_one = i;
+            }
+
+            if i {
+                result = result.mul(cs, self);
+            }
```
same thing here, seems like this will cause divergent circuits based on input

ZamDimon replied:
We are currently using this function for constant exponents, which are the same for all circuits, but I agree that it is unsafe to leave the function further as it is since homogeneity is not guaranteed at this level. I made it homogeneous ;)

jules reviewed (changes requested):
looking great guys, just a few final notes :)

ZamDimon commented:
We moved the `modexp` implementation to [`era-zkevm-circuits`](https://github.com/matter-labs/era-zkevm_circuits/pull/32), as requested. Now, this pull request contains only tower extensions and algebraic torus implementation.

jules reviewed (approved):
all looks good from here!
